### PR TITLE
Add group snapshot controller unit tests

### LINKED ISSUE #1082: Add unit tests for Volume Group Snapshot

### PR DESCRIPTION
**Pull request by leonardoce**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

Adds unit tests for the group snapshot controller.

**Which issue(s) this PR fixes**:

Fixes #1082

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Review by yati1998 — CHANGES_REQUESTED** [review]
I would suggest to add the test for create,delete, update in single file

**Comment by leonardoce** [comment]
> I would suggest to add the test for create,delete, update in single file

Thank you, @yati1998, for having a look.
I split the unit tests because that's what it was done for volume snapshots.
Do you think it would better for volume group snapshots to have a different approach?

File: pkg/common-controller/groupsnapshot_delete_test.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+				map[string]string{
+					"app.kubernetes.io/name": "postgresql",
+				}),
+			initialVolumes: newVolumeCoupleArray("volume6-1", "pv-uid6-1", "pv-handle6-1", "1Gi", "pvc-uid6-1", "claim1-1", v1.VolumeBound, v1.PersistentVolumeReclaimDelete, classGold),
+			errors:         noerrors,
+			test:           testSyncGroupSnapshot,
+			expectSuccess:  true,
+		},
+		{
+			name: "6-3 - dynamic group snapshot have been deleted, retention policy set to delete - volume snapshot content will be deleted",
```
Comment (xing-yang) [inline pkg/common-controller/groupsnapshot_delete_test.go:100]: Add a test for pre-provisioned group snapshot?

File: pkg/common-controller/groupsnapshot_finalizer_test.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2019 The Kubernetes Authors.
```
Comment (xing-yang) [inline pkg/common-controller/groupsnapshot_finalizer_test.go:2]: 2024

File: pkg/common-controller/groupsnapshot_finalizer_test.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+	crdv1 "github.com/kubernetes-csi/external-snapshotter/client/v8/apis/volumesnapshot/v1"
+	"github.com/kubernetes-csi/external-snapshotter/v8/pkg/utils"
+	v1 "k8s.io/api/core/v1"
+)
+
+func TestGroupSnapshotFinalizer(t *testing.T) {
+	tests := []controllerTest{
+		{
+			name: "7-1 - finalizer is added on dynamically provisioned group contents if the storage class specify a Deletion retain policy",
```
Comment (xing-yang) [inline pkg/common-controller/groupsnapshot_finalizer_test.go:30]: Why 7-1?

storage class -> volume group snapshot class?

File: pkg/common-controller/groupsnapshot_finalizer_test.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+				map[string]string{
+					"app.kubernetes.io/name": "postgresql",
+				}),
+			initialVolumes: newVolumeCoupleArray("volume6-1", "pv-uid6-1", "pv-handle6-1", "1Gi", "pvc-uid6-1", "claim1-1", v1.VolumeBound, v1.PersistentVolumeReclaimDelete, classGold),
+			errors:         noerrors,
+			test:           testSyncGroupSnapshot,
+			expectSuccess:  true,
+		},
+		{
+			name: "7-2 - finalizer is not added on dynamically provisioned group contents if the storage class specify the Retain retain policy",
```
Comment (xing-yang) [inline pkg/common-controller/groupsnapshot_finalizer_test.go:69]: storage class -> volume group snapshot class?

File: pkg/common-controller/groupsnapshot_update_test.go [hunk 5]
```diff
@@ -0,0 +1,261 @@
+/*
+Copyright 2019 The Kubernetes Authors.
```
Comment (xing-yang) [inline pkg/common-controller/groupsnapshot_update_test.go:2]: 2024

File: pkg/common-controller/framework_test.go [hunk 6]
```diff
@@ -185,6 +202,35 @@ func isTestError(err error) bool {
...
+		for k, v := range labels {
+			pvcs[i].ObjectMeta.Labels[k] = v
+		}
+	}
+	return pvcs
+}
+
+func withVolumesCSIDriverName(pvs []*v1.PersistentVolume, driverName string) []*v1.PersistentVolume {
+	for i := range pvs {
+		pvs[i].Spec.CSI.Driver = driverName
```
Comment (Madhu-1) [inline pkg/common-controller/framework_test.go:219]: .CSI can be nil as well, can we need to handle that as well

File: pkg/common-controller/framework_test.go [hunk 7]
```diff
@@ -370,24 +534,83 @@ func (r *snapshotReactor) React(action core.Action) (handled bool, ret runtime.O
...
+
+			err = json.Unmarshal(modified, storedGroupSnapshot)
+			if err != nil {
+				return true, nil, err
+			}
+
+			storedVer, _ := strconv.Atoi(storedGroupSnapshot.ResourceVersion)
+			storedGroupSnapshot.ResourceVersion = strconv.Itoa(storedVer + 1)
+		} else {
+			return true, nil, fmt.Errorf("cannot update snapshot %s: snapshot not found", action.GetName())
```
Comment (Madhu-1) [inline pkg/common-controller/framework_test.go:565]:
```suggestion
return true, nil, fmt.Errorf("cannot update group snapshot %s: group snapshot not found", action.GetName())
````

File: pkg/common-controller/framework_test.go [hunk 8]
```diff
@@ -401,17 +624,39 @@ func (r *snapshotReactor) React(action core.Action) (handled bool, ret runtime.O
...
+	case action.Matches("delete", "volumegroupsnapshots"):
+		name := action.(core.DeleteAction).GetName()
+		klog.V(4).Infof("deleted snapshot %s", name)
+		_, found := r.groupSnapshots[name]
+		if found {
+			delete(r.groupSnapshots, name)
+			r.changedSinceLastSync++
+			return true, nil, nil
+		}
+		return true, nil, fmt.Errorf("cannot delete snapshot %s: not found", name)
```
Comment (Madhu-1) [inline pkg/common-controller/framework_test.go:658]:
```suggestion
return true, nil, fmt.Errorf("cannot delete group snapshot %s: not found", name)
```

File: pkg/common-controller/framework_test.go [hunk 9]
```diff
@@ -1053,6 +1487,64 @@ func newSnapshot(
...
+
+	if !nilStatus {
+		groupSnapshot.Status = &crdv1alpha1.VolumeGroupSnapshotStatus{
+			CreationTime: creationTime,
+			ReadyToUse:   readyToUse,
+			Error:        err,
+		}
+	}
+
+	if boundContentName != "" {
```
Comment (Madhu-1) [inline pkg/common-controller/framework_test.go:1522]: we also need `!nilStatus` check here as well right?

File: pkg/common-controller/groupsnapshot_create_test.go [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+	crdv1 "github.com/kubernetes-csi/external-snapshotter/client/v8/apis/volumesnapshot/v1"
+	"github.com/kubernetes-csi/external-snapshotter/v8/pkg/utils"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var groupSnapshotClasses = []*crdv1alpha1.VolumeGroupSnapshotClass{
+	{
+		TypeMeta: metav1.TypeMeta{
+			Kind: "VolumeSnapshotClass",
```
Comment (Madhu-1) [inline pkg/common-controller/groupsnapshot_create_test.go:32]: Shouldn't it be "volumeGroupSnapshotClass"?

File: pkg/common-controller/groupsnapshot_create_test.go [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+				map[string]string{
+					"app.kubernetes.io/name": "postgresql",
+				}),
+			initialVolumes: newVolumeCoupleArray("volume6-1", "pv-uid6-1", "pv-handle6-1", "1Gi", "pvc-uid6-1", "claim1-1", v1.VolumeBound, v1.PersistentVolumeReclaimDelete, classGold),
+			errors:         noerrors,
+			test:           testSyncGroupSnapshot,
+			expectSuccess:  true,
+		},
+		{
+			name: "3-1 - unsuccessful create group snapshot with no existent group snapshot class",
```
[the remark this answers is not in the document]
Comment (leonardoce) [reply pkg/common-controller/groupsnapshot_create_test.go:101]: Good question, but I don't remember the answer. I'm reordering them.

File: pkg/common-controller/groupsnapshot_create_test.go (hunk 10 above)
Comment (leonardoce) [reply pkg/common-controller/groupsnapshot_create_test.go:32]: Yup! Thank you!

File: pkg/common-controller/groupsnapshot_finalizer_test.go (hunk 4 above)
Comment (leonardoce) [reply pkg/common-controller/groupsnapshot_finalizer_test.go:69]: Yes! Thank you!

File: pkg/common-controller/framework_test.go (hunk 6 above)
Comment (leonardoce) [reply pkg/common-controller/framework_test.go:219]: Yes! Good catch, thank you.

File: pkg/common-controller/framework_test.go (hunk 8 above)
Comment (leonardoce) [reply pkg/common-controller/framework_test.go:658]: Fixed, thank you.

File: pkg/common-controller/framework_test.go (hunk 7 above)
Comment (leonardoce) [reply pkg/common-controller/framework_test.go:565]: Fixed, thank you

File: pkg/common-controller/groupsnapshot_create_test.go [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+		Driver:         mockDriverName,
+		Parameters:     class1Parameters,
+		DeletionPolicy: crdv1.VolumeSnapshotContentRetain,
+	},
+}
+
+func TestCreateGroupSnapshotSync(t *testing.T) {
+	tests := []controllerTest{
+		{
+			name: "1-1 - successful dynamically-provisioned create group snapshot with group snapshot class gold",
```
[the remark this answers is not in the document]
Comment (leonardoce) [reply pkg/common-controller/groupsnapshot_create_test.go:70]: The test classification was not precise, I marked the pre-provisioned ones in a clearer way.

File: pkg/common-controller/groupsnapshot_delete_test.go (hunk 1 above)
Comment (leonardoce) [reply pkg/common-controller/groupsnapshot_delete_test.go]: The test classification was not precise, I marked the pre-provisioned ones in a clearer way.

File: pkg/common-controller/framework_test.go [hunk 13]
```diff
@@ -305,6 +406,46 @@ func (r *snapshotReactor) React(action core.Action) (handled bool, ret runtime.O
...
 
+	case action.Matches("patch", "volumegroupsnapshotcontents"):
+		content := &crdv1alpha1.VolumeGroupSnapshotContent{}
+		action := action.(core.PatchAction)
+
+		// Check and bump object version
+		storedGroupSnapshotContent, found := r.groupContents[action.GetName()]
+		if found {
+			// Apply patch
+			storedSnapshotBytes, err := json.Marshal(storedGroupSnapshotContent)
```
Comment (Madhu-1) [inline pkg/common-controller/framework_test.go:417]: storedSnapshotBytes to storedGroupSnapshotBytes

File: pkg/common-controller/framework_test.go [hunk 14]
```diff
@@ -401,17 +627,39 @@ func (r *snapshotReactor) React(action core.Action) (handled bool, ret runtime.O
...
 		if found {
 			delete(r.snapshots, name)
 			r.changedSinceLastSync++
 			return true, nil, nil
 		}
 		return true, nil, fmt.Errorf("cannot delete snapshot %s: not found", name)
 
+	case action.Matches("delete", "volumegroupsnapshots"):
+		name := action.(core.DeleteAction).GetName()
+		klog.V(4).Infof("deleted snapshot %s", name)
```
Comment (Madhu-1) [inline pkg/common-controller/framework_test.go:654]: deleted volume group snapshot

File: pkg/common-controller/framework_test.go [hunk 15]
```diff
@@ -1053,6 +1490,64 @@ func newSnapshot(
 	return &snapshot
 }
 
+func newGroupSnapshot(
+	snapshotName, snapshotUID string, selectors map[string]string, targetContentName, snapshotClassName, boundContentName string,
```
Comment (Madhu-1) [inline pkg/common-controller/framework_test.go:1494]: `snapshotName` to `groupSnapshotName` `snapshotUID` to `groupSnapshotUID` `snapshotClassName` to `groupSnapshotClassName`

File: pkg/common-controller/framework_test.go [hunk 16]
```diff
@@ -1053,6 +1490,64 @@ func newSnapshot(
...
+		groupSnapshot.Spec.Source.VolumeGroupSnapshotContentName = &targetContentName
+	}
+	if withAllFinalizers {
+		return withGroupSnapshotFinalizers([]*crdv1alpha1.VolumeGroupSnapshot{&groupSnapshot}, utils.VolumeGroupSnapshotContentFinalizer, utils.VolumeGroupSnapshotBoundFinalizer)[0]
+	}
+	return &groupSnapshot
+}
+
+func newGroupSnapshotArray(
+	snapshotName, snapshotUID string, selectors map[string]string, targetContentName, snapshotClassName, boundContentName string,
```
Comment (Madhu-1) [inline pkg/common-controller/framework_test.go:1543]: same here as well, can we have the inputs to match the group like groupSnapshotName etc

**Review by Madhu-1 — COMMENTED** [review]
small nits, LGTM

File: pkg/common-controller/framework_test.go (hunk 13 above)
Comment (leonardoce) [reply pkg/common-controller/framework_test.go:417]: Fixed!

File: pkg/common-controller/framework_test.go (hunk 14 above)
Comment (leonardoce) [reply pkg/common-controller/framework_test.go:654]: Fixed!

File: pkg/common-controller/framework_test.go (hunk 15 above)
Comment (leonardoce) [reply pkg/common-controller/framework_test.go:1494]: Fixed!

File: pkg/common-controller/framework_test.go (hunk 16 above)
Comment (leonardoce) [reply pkg/common-controller/framework_test.go:1543]: Fixed! I changed newGroupSnapshotContent(Array) too

**Comment by leonardoce** [comment]
Thank you @Madhu-1!

File: pkg/common-controller/framework_test.go [hunk 17]
```diff
@@ -924,6 +1277,73 @@ func newContent(contentName, boundToSnapshotUID, boundToSnapshotName, snapshotHa
 	return &content
 }
 
+func newGroupSnapshotContent(groupSnapshotContentName, boundToSnapshotUID, boundToGroupSnapshotName, groupSnapshotHandle, groupSnapshotClassName string, desiredVolumeHandles []string, volumeGroupHandle string,
```
Comment (xing-yang) [inline pkg/common-controller/framework_test.go:1280]: boundToSnapshotUID -> boundToGroupSnapshotUID

File: pkg/common-controller/groupsnapshot_create_test.go [hunk 18]
```diff
@@ -0,0 +1,377 @@
...
+				"test.csi.driver.name",
+			),
+			errors:        noerrors,
+			test:          testSyncGroupSnapshot,
+			expectSuccess: false,
+		},
+		{
+			name: "1-8 - successful pre-provisioned group snapshot",
+			initialGroupSnapshots: newGroupSnapshotArray(
+				"snap-1-1", "snapuid1-1", map[string]string{
```
Comment (xing-yang) [inline pkg/common-controller/groupsnapshot_create_test.go:273]: "snap-1-1", "snapuid1-1" are group snapshot name and group snapshot uuid?  Can you name them accordingly?  If it is a group snapshot, please add "group" to the name to differentiate it from individual ones. This is a general comment.

**Comment by leonardoce** [comment]
Done @xing-yang, and thanks for the advice.
I also added another test for pre-provisioned volume group snapshots, testing the mismatch between a pre-provisioned volume group snapshot and a dynamically-provisioned volume group snapshot content.

**Pull request merged**